hips/hip-0020.md

```diff
@@ -0,0 +1,133 @@
+
+The design specified here is meant to enable the current charts to work as expected while providing
+space for more radical changes.
+
+## Specification
+
+Describe the syntax and semantics of any new feature.
+
+Helm has numerous packages that do various things, from working with charts to storing release
+information. These packages all expect there to be one version of a particular thing. To support
+multiple versions of charts, the `chart`, `chartutil`, `engine`, and `release` packages will be made
+multi-version. (e.g., engine will have a v1 and v2 versions). In addition to enabling a new version of
+charts, this will enable the gotpl engine to evolve as needed to support the new version of charts. 
+The `chartutil` may be combined with the `chart` package and the `releaseutil` package may be
+combined with the `release` package for simplification of the package structure.
+
+The versioning of the packages will follow the same structure that Kubernetes does with its APIs.
+This means a thing will have a directory and within it will be sub-directories for the versions. The
+version specific implementation will be in the version specific sub-directory. For example,
+
+```
+chart
+├── v2
+└── v3
+```
+
+The version will not use a separate Git repository or Go modules. The reason for this is the added
+complexity of managing the repositories and modules is added work to manage changes which will
+slow down velocity, will make releases more complex, and make the Helm SDK more complicated
+to work with.
+
+The existing versions of these packages will be externally facing while the new versions will be
+developed in `internal` as experiments until they are complete enough for release. When ready for
+release, these packages will be moved to the public locations.
+
+To illustrate this, while in development the `chart` package will have the following structure:
+
+```
+internal/chart
+└── v3
+pkg/chart
+└── v2
+```
+
+Once the new chart packages have (a) a stable API and (b) are feature complete the structure will
+be merged to look like:
+
+```
+pkg/chart
+├── v2
+└── v3
+```
+
+While in development, as an experiment, the `pkg/gates` package will be used to create an opt-in
+environment variable to enable a new chart version. This is how the OCI experimental feature was
+handled.
+
+## Backwards compatibility
+
+This development and process is designed with backwards compatibility in mind. The package
+locations will change, which is ok in a major version of Helm. The existing charts will be preserved
+so that their installation process continues to work as expected from Helm v3. New features and
+changes to charts can be added in without impacting existing charts.
+
+While the new chart version is being developed and is going through breaking changes, the chart
+will be an opt-in feature which will enable us to warn users about the state of it.
+
+## Security implications
+
+N/A
+
+## How to teach this
+
+There are a few ways to teach this:
```

Comment:
Should we also teach users how to migrate?

I feel like we also want a section on migrating charts even if migrating is merely changing the charts `apiVersion` to `v3`.

Reply:
I think that can be spelled out directly in the documentation, etc.